packages/react-instantsearch-hooks/src/hooks/__tests__/useInstantSearch.test.tsx

```diff
@@ -0,0 +1,458 @@
+            }),
+        }
+      );
+
+      // augmented once middleware was used
+      expect(result.current.middleware).toEqual([
+        {
+          creator: middleware,
+          instance: {
+            onStateChange,
+            subscribe,
+            unsubscribe,
+          },
+        },
+      ]);
+
+      expect(subscribe).toHaveBeenCalledTimes(1);
+      expect(onStateChange).toHaveBeenCalledTimes(0);
+      expect(unsubscribe).toHaveBeenCalledTimes(0);
+      expect(middleware).toHaveBeenCalledTimes(1);
+
+      await waitForNextUpdate();
+
+      expect(subscribe).toHaveBeenCalledTimes(1);
+      expect(onStateChange).toHaveBeenCalledTimes(0);
+      expect(unsubscribe).toHaveBeenCalledTimes(0);
+
+      // simulate a change in query
+      result.current.renderState.indexName.searchBox!.refine('new query');
+      await waitForNextUpdate();
+
+      expect(subscribe).toHaveBeenCalledTimes(1);
+      expect(onStateChange).toHaveBeenCalledTimes(1);
+      expect(unsubscribe).toHaveBeenCalledTimes(0);
+
+      unmount();
+
+      expect(subscribe).toHaveBeenCalledTimes(1);
+      expect(onStateChange).toHaveBeenCalledTimes(1);
+      // unsubscribe is first called by the parent InstantSearch unmounting
+      // 🚨 dispose doesn't remove middleware (because otherwise routing would break)
+      // then unuse is called by this widget itself unmounting.
+      // if only the component with useInstantSearch is unmounted, unsubscribe is called once.
+      expect(unsubscribe).toHaveBeenCalledTimes(2);
+
+      // middleware was removed
+      expect(result.current.middleware).toEqual([]);
+    });
+  });
+
+  describe('refresh', () => {
+    test('refreshes the search', async () => {
+      const searchClient = createSearchClient({});
+      function App() {
+        const { refresh } = useInstantSearch();
+
+        return (
+          <>
+            <button
+              type="button"
+              onClick={() => {
+                refresh();
+              }}
+            >
+              refresh
+            </button>
+          </>
+        );
+      }
+
+      const { container } = render(
+        <InstantSearchHooksTestWrapper searchClient={searchClient}>
+          <App />
+        </InstantSearchHooksTestWrapper>
+      );
+
+      await wait(0);
+
+      expect(searchClient.search).toHaveBeenCalledTimes(1);
+
+      userEvent.click(container.querySelector('button')!);
+
+      expect(searchClient.search).toHaveBeenCalledTimes(2);
+    });
+
+    test('provides a stable reference', async () => {
+      const wrapper = createInstantSearchTestWrapper();
+      const { result, waitForNextUpdate, rerender } = renderHook(
+        () => useInstantSearch(),
+        { wrapper }
+      );
+
+      expect(result.current.refresh).toEqual(expect.any(Function));
+
+      const ref = result.current.refresh;
+
+      await waitForNextUpdate();
+
+      // reference has not changed
+      expect(result.current.refresh).toEqual(ref);
```

Comment:
We want to test the reference equality.

```suggestion
      expect(result.current.refresh).toBe(ref);
```